File: android/core/src/main/kotlin/dev/sn/core/Agent.kt
package dev.sn.core

import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive

/** What the agent is doing, as it does it. The UI renders these. */
sealed interface AgentEvent {
    /** A fragment of the answer. */
    data class Delta(val text: String) : AgentEvent

    data class ToolStarted(val tool: String, val arguments: JsonObject) : AgentEvent

    data class ToolFinished(
        val tool: String,
        val arguments: JsonObject,
        val summary: String,
        val failed: Boolean,
    ) : AgentEvent

    /** The user declined a confirmation prompt. */
    data class ToolDenied(val tool: String, val arguments: JsonObject) : AgentEvent

    /**
     * The turn finished. [newMessages] is everything generated this turn — the
     * user message, assistant turns and tool results — ready to be persisted.
     */
    data class Final(val text: String, val newMessages: List<ChatMessage>) : AgentEvent

    /** The turn could not complete. Already phrased for the user. */
    data class Failed(val message: String, val newMessages: List<ChatMessage>) : AgentEvent
}

data class AgentConfig(
    val systemPrompt: String = DEFAULT_SYSTEM_PROMPT,
    /** How many tool round trips before giving up on a question. */
    val maxSteps: Int = 8,
    /** Tool names that must be confirmed, on top of those marked consequential. */
    val confirmTools: Set<String> = emptySet(),
    /** Tool names switched off entirely. The model is not told they exist. */
    val disabledTools: Set<String> = emptySet(),
)

/** Asked before a consequential tool runs. Returning false declines it. */
fun interface Confirmer {
    suspend fun confirm(tool: Tool, arguments: JsonObject): Boolean
}

/** Records what a tool did, for the audit log. */
fun interface ToolAuditor {
    suspend fun record(
        tool: String,
        arguments: JsonObject,
        decision: String,
        summary: String,
        failed: Boolean,
    )
}

/**
 * The loop: ask the model, run the tools it asks for, repeat until it answers.
 *
 * Emits [AgentEvent]s rather than touching any UI, which is what keeps this
 * module free of Android and testable on a plain JVM.
 */
class Agent(
    private val client: ChatBackend,
    private val registry: ToolRegistry,
    private val config: AgentConfig,
    private val confirmer: Confirmer = Confirmer { _, _ -> true },
    private val auditor: ToolAuditor = ToolAuditor { _, _, _, _, _ -> },
) {

    fun run(prompt: String, history: List<ChatMessage> = emptyList()): Flow<AgentEvent> = flow {
        val userMessage = ChatMessage.user(prompt)
        val newMessages = mutableListOf(userMessage)

        val conversation = mutableListOf<ChatMessage>().apply {
            add(ChatMessage.system(config.systemPrompt))
            addAll(history)
            add(userMessage)
        }
        val specs = registry.specs(config.disabledTools)

        repeat(config.maxSteps) {
            var complete: ChatMessage? = null
            try {
                client.chat(conversation, specs).collect { event ->
                    when (event) {
                        is ChatStream.Delta -> emit(AgentEvent.Delta(event.text))
                        is ChatStream.Complete -> complete = event.message
                    }
                }
            } catch (e: CancellationException) {
                throw e
            } catch (e: OllamaException) {
                emit(AgentEvent.Failed(e.message ?: "Could not reach the model.", newMessages))
                return@flow
            }

            val reply = complete ?: ChatMessage(Role.ASSISTANT, "")
            conversation += reply
            newMessages += reply

            val calls = reply.toolCalls.orEmpty()
            if (calls.isEmpty()) {
                emit(AgentEvent.Final(reply.content, newMessages))
                return@flow
            }

            for (call in calls) {
                val name = call.function.name
                val requested = call.function.arguments
                val tool = registry.find(name, config.disabledTools)

                if (tool != null && needsConfirmation(tool)) {
                    if (!confirmer.confirm(tool, requested)) {
                        emit(AgentEvent.ToolDenied(name, requested))
                        auditor.record(name, requested, "denied", "declined by user", false)
                        val note = ChatMessage.tool(
                            name,
                            "The user declined to run this tool. Do not try it again. " +
                                "Acknowledge briefly and ask what they would prefer.",
                        )
                        conversation += note
                        newMessages += note
                        continue
                    }
                }

                val arguments = tool?.let { coerceArguments(it.parameters, requested) } ?: requested
                emit(AgentEvent.ToolStarted(name, arguments))

                val outcome = execute(tool, name, arguments)
                auditor.record(name, arguments, "ran", outcome.summary, outcome.failed)

                val result = ChatMessage.tool(name, outcome.payload)
                conversation += result
                newMessages += result
                emit(AgentEvent.ToolFinished(name, arguments, outcome.summary, outcome.failed))
            }
        }

        emit(
            AgentEvent.Failed(
                "Stopped after ${config.maxSteps} tool steps without an answer. " +
                    "Raise the step limit in settings, or ask something narrower.",
                newMessages,
            ),
        )
    }

    private fun needsConfirmation(tool: Tool): Boolean =
        tool.consequential || tool.name in config.confirmTools

    private data class Outcome(val payload: String, val summary: String, val failed: Boolean)

    private suspend fun execute(tool: Tool?, name: String, arguments: JsonObject): Outcome {
        if (tool == null) {
            val known = registry.enabled(config.disabledTools).joinToString(", ") { it.name }
            return Outcome(
                "Error: No tool named '$name'. Available tools: $known",
                "no such tool",
                failed = true,
            )
        }
        return try {
            val text = tool.call(arguments).toString()
            Outcome(text, summarize(text), failed = false)
        } catch (e: CancellationException) {
            throw e
        } catch (e: ToolException) {
            Outcome("Error: ${e.message}", e.message ?: "failed", failed = true)
        } catch (e: Exception) {
            // A crashing tool must never lose the conversation. Hand the model
            // the failure and let it decide what to tell the user.
            val detail = "${e::class.simpleName}: ${e.message}"
            Outcome("Error: $name failed unexpectedly: $detail", detail, failed = true)
        }
    }

    private fun summarize(text: String, width: Int = 140): String =
        if (text.length <= width) text else text.take(width) + "…"

    companion object {
        /** Formats tool arguments for a confirmation dialog. */
        fun describeArguments(arguments: JsonObject): List<Pair<String, String>> =
            arguments.map { (key, value) ->
                key to when (value) {
                    is JsonPrimitive -> if (value.isString) value.content else value.toString()
                    else -> value.toString()
                }
            }
    }
}

val DEFAULT_SYSTEM_PROMPT = """
You are sn, a personal assistant running directly on the user's Android phone (a
Samsung Galaxy S23 Ultra). You reach the phone's messages, contacts, calendar,
notifications, files, camera, location and device state through tools.

How to behave:
- Use tools to find things out. Do not guess at the contents of the phone, and
  never invent a phone number, contact, message, event or file path.
- Prefer one targeted tool call over several broad ones. Read before you write.
- Tools that send messages, place calls or take photos are confirmed by the user
  before they run. Call them normally; do not ask for permission in prose first,
  and do not claim something was sent until a tool result says it was.
- If a tool returns an error, tell the user what actually failed rather than
  retrying the same call unchanged.
- The user is on a phone screen. Answer in a few short sentences. No preamble, no
  restating the question, no markdown tables.
""".trimIndent()

File: sn/agent.py
"""The agent loop: ask the model, run the tools it asks for, repeat.

Emits events as it goes so the CLI can stream text and show tool activity
without the loop knowing anything about terminals.
"""

from __future__ import annotations

import json
from dataclasses import dataclass, field
from typing import Any, Callable, Iterator

from .config import Config
from .llm import LLMError, Message, OllamaClient
from .session import Store
from .tools import Toolbox
from .tools.base import Tool, ToolError

# Asked whether a consequential tool may run. Returning False declines it and
# tells the model so, which is different from the tool failing.
Confirmer = Callable[[Tool, dict], bool]


def allow_all(tool: Tool, arguments: dict) -> bool:
    return True


def deny_all(tool: Tool, arguments: dict) -> bool:
    return False


@dataclass
class Event:
    kind: str  # content | tool_start | tool_result | tool_denied | final | error
    text: str = ""
    tool: str = ""
    arguments: dict = field(default_factory=dict)
    result: Any = None


class Agent:
    def __init__(
        self,
        config: Config,
        client: OllamaClient,
        toolbox: Toolbox,
        store: Store | None = None,
        confirm: Confirmer = allow_all,
    ):
        self.config = config
        self.client = client
        self.toolbox = toolbox
        self.store = store
        self.confirm = confirm

    # -- helpers ----------------------------------------------------------

    def _needs_confirmation(self, tool: Tool) -> bool:
        return tool.name in self.config.tools.confirm or tool.consequential

    @staticmethod
    def _arguments(call: dict) -> dict:
        """Pull arguments out of a tool call, whether dict or JSON string."""
        raw = (call.get("function") or {}).get("arguments", {})
        if isinstance(raw, str):
            try:
                raw = json.loads(raw or "{}")
            except json.JSONDecodeError:
                return {}
        return raw if isinstance(raw, dict) else {}

    def _execute(self, name: str, arguments: dict) -> tuple[Any, str | None]:
        """Run one tool. Returns (result, error_message)."""
        try:
            tool = self.toolbox.get(name)
        except ToolError as exc:
            return None, str(exc)

        # Drop arguments the tool does not accept rather than crashing on a
        # model that invents one.
        accepted = set(tool.parameters.get("properties", {}))
        cleaned = {k: v for k, v in arguments.items() if k in accepted}

        try:
            return tool(**cleaned), None
        except ToolError as exc:
            return None, str(exc)
        except TypeError as exc:
            return None, f"Wrong arguments for {name}: {exc}"
        except Exception as exc:  # noqa: BLE001 — a tool must never kill the loop
            return None, f"{name} failed unexpectedly: {type(exc).__name__}: {exc}"

    def _record(self, session_id: int | None, message: dict) -> None:
        if self.store and session_id is not None:
            self.store.append(session_id, message)

    # -- the loop ---------------------------------------------------------

    def run(self, prompt: str, session_id: int | None = None) -> Iterator[Event]:
        history: list[dict] = []
        if self.store and session_id is not None:
            history = self.store.history(session_id, self.config.agent.history_turns * 2)

        user_message = {"role": "user", "content": prompt}
        self._record(session_id, user_message)

        messages: list[dict] = [
            {"role": "system", "content": self.config.agent.system_prompt},
            *history,
            user_message,
        ]
        specs = self.toolbox.specs()

        for step in range(self.config.agent.max_steps):
            reply = Message()
            try:
                for kind, value in self.client.stream_chat(messages, specs):
                    if kind == "content":
                        yield Event("content", text=value)
                    else:
                        reply = value
            except LLMError as exc:
                yield Event("error", text=str(exc))
                return

            assistant_message = reply.to_dict()
            messages.append(assistant_message)
            self._record(session_id, assistant_message)

            if not reply.tool_calls:
                yield Event("final", text=reply.content)
                return

            for call in reply.tool_calls:
                name = (call.get("function") or {}).get("name", "")
                arguments = self._arguments(call)
                tool = self.toolbox.tools.get(name)

                if tool is not None and self._needs_confirmation(tool):
                    if not self.confirm(tool, arguments):
                        yield Event("tool_denied", tool=name, arguments=arguments)
                        note = (
                            "The user declined to run this tool. Do not try it again. "
                            "Acknowledge briefly and ask what they would prefer."
                        )
                        self._append_tool_result(messages, session_id, name, note)
                        if self.store and session_id is not None:
                            self.store.audit(
                                session_id=session_id,
                                tool=name,
                                arguments=arguments,
                                decision="denied",
                            )
                        continue

                yield Event("tool_start", tool=name, arguments=arguments)
                result, error = self._execute(name, arguments)

                if self.store and session_id is not None:
                    self.store.audit(
                        session_id=session_id,
                        tool=name,
                        arguments=arguments,
                        decision="ran",
                        result=result,
                        error=error,
                    )

                payload = f"Error: {error}" if error else json.dumps(result, default=str)
                self._append_tool_result(messages, session_id, name, payload)
                yield Event(
                    "tool_result", tool=name, arguments=arguments, result=error or result
                )

        yield Event(
            "error",
            text=(
                f"Stopped after {self.config.agent.max_steps} tool steps without a final "
                "answer. Raise agent.max_steps in the config, or ask something narrower."
            ),
        )

    def _append_tool_result(
        self, messages: list[dict], session_id: int | None, name: str, content: str
    ) -> None:
        message = {"role": "tool", "tool_name": name, "content": content}
        messages.append(message)
        self._record(session_id, message)
